File: Python3/프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.py
def solution(number, k):
    stack = []
    
    for num in number:
        # 스택의 최상단이 num보다 작을 경우
        while stack and k > 0 and stack[-1] < num:
            stack.pop() # 제거
            k -= 1 # 삭제 횟수 차감
        stack.append(num) # 스택에 추가
    
    # k가 아직 남아있는 경우 내림차순으로 뒤에를 제거
    if k > 0:
        stack = stack[:-k]
        
    return ''.join(stack)

File: C++/프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.cpp
#include <string>
#include <stack>

using namespace std;

string solution(string number, int k) {
    string answer = "";
    
    for (int i = 0; i < number.size(); i++) {
        // 제거할 숫자가 남아있고, 현재 숫자가 앞에 있는 숫자보다 크면 앞의 숫자를 제거
        while ((!answer.empty()) && (k > 0) && (answer.back() < number[i])) {
            answer.pop_back();
            k--;  // 제거할 숫자 감소
        }
        
        // 결과에 숫자 추가
        answer.push_back(number[i]);
    }
    
    // 만약 k가 남아있는 경우, 뒤에서부터 k개를 자름
    if (k > 0) {
        answer = answer.substr(0, answer.size() - k);
    }

    return answer;
}
